code/core/system/proxy-velodyne16/src/velodyne16Decoder.cpp

```diff
@@ -270,48 +299,87 @@ void Velodyne16Decoder::nextString(const string &payload) {
-                            m_segment[m_startID + 1] = yData;
-                            m_segment[m_startID + 2] = zData;
-                            m_segment[m_startID + 3] = intensity;
+                            } else {//distance+azimuth+vertical angle+intensity
+                                m_segment[m_startID] = m_distance;
+                                m_segment[m_startID + 1] = m_currentAzimuth;
+                                m_segment[m_startID + 2] = m_verticalAngle[sensorID];
+                                m_segment[m_startID + 3] = (float)thirdByte;//intensity; 
+                            }
                             m_pointIndexSPC++;
                             m_startID += m_NUMBER_OF_COMPONENTS_PER_POINT;
-                        }
+                        }   
                     }
                     
-                    if(m_withCPC && m_pointIndexCPC < m_MAX_POINT_SIZE){
-                        //Store distance in cm in an array of uint16_t type
-                        m_16Sensors[sensorID] = ntohs(firstByte * 256 + secondByte)/5;   
-                        if(sensorID==15){
-                            for(uint8_t index=0;index<16;index++){
-                                m_distanceStringStream.write((char*)(&m_16Sensors[m_sensorOrderIndex[index]]),2);
+                    if (m_withCPC && m_pointIndexCPC < m_MAX_POINT_SIZE) {
+                        if (m_CPCIntensityOption == 0 || m_CPCIntensityOption == 2) {
+                            //Store distance with resolution 2mm in an array of uint16_t type
+                            m_16SensorsNoIntensity[sensorID] = ntohs(firstByte * 256 + secondByte);
+                            if (m_distanceEncoding == 0) {
+                                //Store distance with resolution 1cm in an array of uint16_t type
+                                m_16SensorsNoIntensity[sensorID] = m_16SensorsNoIntensity[sensorID] / 5;  
+                            }
+                            
+                            if (sensorID == 15) {
+                                for (uint8_t index = 0; index < 16; index++) {
+                                    m_distanceStringStreamNoIntensity.write((char*)(&m_16SensorsNoIntensity[m_sensorOrderIndex[index]]),2);
+                                }
+                            }
+                            m_pointIndexCPC++;    
+                        }
+                        
+                        if (m_CPCIntensityOption == 1 || m_CPCIntensityOption == 2) {
+                            //Store distance with resolution 2mm in an array of uint16_t type
+                            uint16_t distance = ntohs(firstByte * 256 + secondByte);
+                            if (m_distanceEncoding == 0) {
+                                distance = distance / 5;
+                            }
+                            distance = distance >> m_numberOfBitsForIntensity; ////Reserve the upper n bits for intensity
```

Comment:
This shift is an error as we modify the precision here. You should not shift the value by numberOfIntensityBits here but mask the upper numberOfIntensityBits to zero. Thereby, we are able to encode both